### PR TITLE
[SPARK-21506][DOC]The description of "spark.executor.cores" may be not  correct

### PR DESCRIPTION
**Pull request by 10110346**
## What changes were proposed in this pull request?

The number of cores assigned to each executor is configurable. When this is not explicitly set,  multiple executors from the same application may be launched on the same worker too.

## How was this patch tested?
N/A

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 1]
```diff
@@ -580,7 +580,9 @@ private[deploy] class Master(
    * The number of cores assigned to each executor is configurable. When this is explicitly set,
    * multiple executors from the same application may be launched on the same worker if the worker
    * has enough cores and memory. Otherwise, each executor grabs all the cores available on the
-   * worker by default, in which case only one executor may be launched on each worker.
+   * worker by default, in which case only one executor may be launched on each worker, but if it
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: I think that's a bit of a corner case to cover. The text already suggests that, _usually_, only one schedules.

Comment (10110346) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: thanks, The description in " configuration.md"  i think should be modified

**Comment by srowen** [comment]
I'm not sure this is an improvement, though not really against it.

**Comment by jiangxb1987** [comment]
AFAIK when `spark.executor.cores` is not set, we would keep assigning free cores to the executor until `app.coresLeft` is zero or there is no more free cores left. So I'm not convinced the comment should be updated. Please correct me if I'm wrong, thanks!

**Comment by 10110346** [comment]
@jiangxb1987  If `app.coresLeft` is not  zero and  there is no more free cores left, it is not ending.Waiting for some workers have  free cores, this app will be  assigned cores continue, and this time another executor may be launched on the same worker.

**Comment by jiangxb1987** [comment]
How can we wait for new free cores on the same worker, if we only start one executor for each worker initially?

**Comment by 10110346** [comment]
If the other application is finished, it will release cores

**Comment by jiangxb1987** [comment]
yea, I understand the issue now. Thank you for clarify!

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 2]
```diff
@@ -580,7 +580,8 @@ private[deploy] class Master(
    * The number of cores assigned to each executor is configurable. When this is explicitly set,
    * multiple executors from the same application may be launched on the same worker if the worker
    * has enough cores and memory. Otherwise, each executor grabs all the cores available on the
-   * worker by default, in which case only one executor may be launched on each worker.
+   * worker by default, in which case only one executor may be launched on each worker only one
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Perhaps rephrase this to:
```
in which case only one executor may be launched on each worker during one schedule iteration.
```
?

File: docs/configuration.md [hunk 3]
```diff
@@ -1106,7 +1106,7 @@ Apart from these, the following properties are also available, and may be useful
     parameter allows an application to run multiple executors on the
     same worker, provided that there are enough cores on that
     worker. Otherwise, only one executor per application will run on
-    each worker.
+    each worker usually.
```
Comment (jiangxb1987) [inline docs/configuration.md]:
```
    In standalone and Mesos coarse-grained modes, setting this
    parameter allows an application to launch multiple executors on the
    same worker in one schedule iteration, provided that there are enough
    cores on that worker. Otherwise, only one executor per application
    will be scheduled on each worker during one schedule iteration.
```

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 4]
```diff
@@ -580,7 +580,8 @@ private[deploy] class Master(
...
+   * schedules.
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Should also add:
```
Note that when `spark.executor.cores` is not set, we may still launch multiple executors from the same application on the same worker. Consider appA and appB both have one executor running on worker1, and appA.coresLeft > 0, then appB is finished and release all its cores on worker1, thus for the next schedule iteration, appA launchs a new executor that grabs all the free cores on worker1, therefore we get mulfiple executors from appA running on worker1.
```

File: docs/configuration.md (hunk 3 above)
Comment (10110346) [inline docs/configuration.md]: thanks, I will update it

**Comment by 10110346** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 5]
```diff
@@ -580,7 +580,13 @@ private[deploy] class Master(
...
    * multiple executors from the same application may be launched on the same worker if the worker
    * has enough cores and memory. Otherwise, each executor grabs all the cores available on the
-   * worker by default, in which case only one executor may be launched on each worker.
+   * worker by default, in which case only one executor may be launched on each worker during one
+   * single schedule iteration.
+   * Note that when `spark.executor.cores` is not set, we may still launch multiple executors from
+   * the same application on the same worker. Consider appA and appB both have one executor running
+   * on worker1, and appA.coresLeft > 0, then appB is finished and release all its cores on worker1,
+   * thus for the next schedule iteration, appA launchs a new executor that grabs all the free cores
+   * on worker1, therefore we get mulfiple executors from appA running on worker1.
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: nit: multiple.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 6]
```diff
@@ -580,7 +580,13 @@ private[deploy] class Master(
    * The number of cores assigned to each executor is configurable. When this is explicitly set,
    * multiple executors from the same application may be launched on the same worker if the worker
    * has enough cores and memory. Otherwise, each executor grabs all the cores available on the
-   * worker by default, in which case only one executor may be launched on each worker.
+   * worker by default, in which case only one executor may be launched on each worker during one
+   * single schedule iteration.
+   * Note that when `spark.executor.cores` is not set, we may still launch multiple executors from
+   * the same application on the same worker. Consider appA and appB both have one executor running
+   * on worker1, and appA.coresLeft > 0, then appB is finished and release all its cores on worker1,
+   * thus for the next schedule iteration, appA launchs a new executor that grabs all the free cores
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: nit: launches

File: docs/configuration.md [hunk 7]
```diff
@@ -1103,10 +1103,10 @@ Apart from these, the following properties are also available, and may be useful
...
 
     In standalone and Mesos coarse-grained modes, setting this
-    parameter allows an application to run multiple executors on the
-    same worker, provided that there are enough cores on that
-    worker. Otherwise, only one executor per application will run on
-    each worker.
+    parameter allows an application to launch multiple executors on the
+    same worker in one single schedule iteration, provided that there are enough
+    cores on that worker. Otherwise, only one executor per application
+    will be scheduled on each worker during one single schedule iteration.
```
Comment (jerryshao) [inline docs/configuration.md]: My hunch is that user may not know the actual meaning of "one single schedule iteration" even if you're trying to clarify the behavior here, unless users know the implementation here.

Comment (10110346) [inline docs/configuration.md]: Thanks. I agree with you.
How to describe it will be better?

Comment (jerryshao) [inline docs/configuration.md]: I'm not sure how to address it better, but my thinking is that instead of trying to address it clearly with limited words, leave the corner case as not to say might be better. Just my two cents.

File: docs/configuration.md [hunk 8]
```diff
@@ -1106,7 +1106,7 @@ Apart from these, the following properties are also available, and may be useful
     parameter allows an application to run multiple executors on the
     same worker, provided that there are enough cores on that
     worker. Otherwise, only one executor per application will run on
-    each worker.
+    each worker in most cases.
```
Comment (jiangxb1987) [inline docs/configuration.md]: I think we should describe when we will break this assumption, either we treat it a bug and try to fix that, or we explicitly document the behavior. I don't think it makes much sense to just add some vague restriction such as "in most cases".

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 9]
```diff
@@ -580,7 +580,13 @@ private[deploy] class Master(
    * The number of cores assigned to each executor is configurable. When this is explicitly set,
    * multiple executors from the same application may be launched on the same worker if the worker
    * has enough cores and memory. Otherwise, each executor grabs all the cores available on the
-   * worker by default, in which case only one executor may be launched on each worker.
+   * worker by default, in which case only one executor may be launched on each worker during one
```
Comment (umehrot2) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Shouldn't this line be:
worker by default, in which case only one executor per application may be launched on each worker during one

**Review by umehrot2 — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala (hunk 9 above)
Comment (10110346) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: ok,thanks

**Comment by srowen** [comment]
I don't object to this but I suppose I'm not clear if this was a point of confusion that needed a fix.

**Comment by jiangxb1987** [comment]
In the past days we often assume that in standalone mode, if we don't set the "spark.executor.cores" property, then we should have no more than one executor per host, but this PR explains a senario that we could still have multiple executors running on the same host under such condition, I think it would be helpful for us to understand some corner case behaviors, so yea I think we should document the behavior somewhere. Also cc @JoshRosen @cloud-fan

**Comment by 10110346** [comment]
@JoshRosen @cloud-fan  Could you help to review it ? thanks

**Comment by jiangxb1987** [comment]
@10110346 Could you update the PR as @jerryshao suggested? Thanks!

**Comment by cloud-fan** [comment]
LGTM, merging to master!
